### DISCUSSION TITLE
How to create custom observables?

### DISCUSSION THREAD

**Discussion by hummingly** [solidjs/solid#676] [Q&A] [answered]
Hello everybody!

Recently, I ported a very small web app prototype from React + MobX to Solid. It's nice that I get reactivity out of the box but Solidjs does not track changes of fields in arrays of objects in stores well.

In my case I had an object defined as:

```ts
type Frame = { buffer: Uint8ClampedArray, image: ImageData };
```

Changes to the byte buffer are not tracked at all eventhough I thought a typed array index property is the same as an array index. Then I added a dirty field which is of type boolean but setting does not notify subscribers.

```ts
setState("frames", frameIndex, "dirty", v => !v);

// In component somewhere...
createEffect(() => {
  // Does not trigger effect
  const _ = state.frames[state.selection].dirty;

  // Update canvas...
};
```

I solved it by separating the dirty field into a separate array in the store. Changing the value of a boolean array does trigger a change.

```ts
// This works!
setState("dirtyFrames", frameIndex, v => !v);
```

However, it is annoying that I have to do all operations twice when I muatate the frames array.

Anyways it would be nice to learn how to build our own primitives. In MobX there is an atom object that is like a boolean that can propagate changes.

**Comment by atk** [comment]
As the documentation states (https://www.solidjs.com/docs/latest/api#updating-stores), arrays need to be replaced to be reactively tracked, but there's also the index form of setState that you could use:

```typescript
setState("frames", [frameIndex], (frame) => frame.dirty = !frame.dirty);
```

**Reply by hummingly** [reply]
Even with index form, it does not react but thanks  for the suggestion. I found another solution though I wonder if it should be just documented or fixed.

**Comment by hummingly** [comment, marked as answer]
The solution was to not use a class but instead an object. Then I wonder if that is a bug or supposed to be  like this.

```ts
class Frame {
  dirty: boolean = false; // Will not be tracked when accessed in effect.
  ...
}

type Frame = { dirty: boolean; ... } // This will be tracked on access!
```

**Reply by ryansolid** [reply]
It is intentional. We don't wrap Classes, only plain objects and arrays, and other primitives values. This prevents wrapping stuff like HTMLElements or Regex or Dates. This is something I borrowed from MobX which has the same restriction. Did you ever have this working in MobX? And if so how did you approach it?

**Reply by hummingly** [reply]
The MobX docs encourage the usage of classes for stores and entities. In a bigger project I also use decorators to annotate which property should be reactive  and which method should be an action. In the constructor `this` is passed to the `makeObservable` function. [Example in documentation without decorators](https://mobx.js.org/observable-state.html)

**Comment by ryansolid** [comment]
I see. While I did get the whether to wrap code directly from MobX there are ways to get around it. They even say they don't wrap nested classes automatically in a note under their observable helper in the docs.
Technically in Solid you can pass a class to `createStore` top-level just not nested. To nest classes you'd need to nest proxies created by `createStore`.
